<a href="https://colab.research.google.com/github/sjw8793/DS2023-HateCrimeAnalysis/blob/main/Hypothesis_test(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
from google.colab import drive
drive.mount('/content/gdrive')
%cd "/content/gdrive/MyDrive/SSU2023_1/DataScience/dataset"

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).
/content/gdrive/MyDrive/SSU2023_1/DataScience/dataset


In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import random

## Preprocessing - 데이터 합치기
### 인구수 기반으로 정규화한 데이터 & 주별 소득 & 주별 인종 비율

In [6]:
norm_pop_2017 = pd.read_csv('normalized_dataset_2017.csv')
norm_pop_2018 = pd.read_csv('normalized_dataset_2018.csv')
norm_pop_2019 = pd.read_csv('normalized_dataset_2019.csv')
norm_pop_2020 = pd.read_csv('normalized_dataset_2020.csv')
norm_pop_2021 = pd.read_csv('normalized_dataset_2021.csv')

income_2017 = pd.read_csv('income2017.csv')
income_2018 = pd.read_csv('income2018.csv')
income_2019 = pd.read_csv('income2019.csv')
income_2020 = pd.read_csv('income2020.csv')
income_2021 = pd.read_csv('income2021.csv')

df1_2017 = pd.merge(norm_pop_2017, income_2017, on='state_name')
df1_2018 = pd.merge(norm_pop_2018, income_2018, on='state_name')
df1_2019 = pd.merge(norm_pop_2019, income_2019, on='state_name')
df1_2020 = pd.merge(norm_pop_2020, income_2020, on='state_name')
df1_2021 = pd.merge(norm_pop_2021, income_2021, on='state_name')

df1_2017 = df1_2017.drop(['number_of_victims_normalized','number_of_offenders_normalized'], axis=1)
df1_2018 = df1_2018.drop(['number_of_victims_normalized','number_of_offenders_normalized'], axis=1)
df1_2019 = df1_2019.drop(['number_of_victims_normalized','number_of_offenders_normalized'], axis=1)
df1_2020 = df1_2020.drop(['number_of_victims_normalized','number_of_offenders_normalized'], axis=1)
df1_2021 = df1_2021.drop(['number_of_victims_normalized','number_of_offenders_normalized'], axis=1)

norm_race_2017 = pd.read_csv('norm_2017_race.csv')
norm_race_2018 = pd.read_csv('norm_2018_race.csv')
norm_race_2019 = pd.read_csv('norm_2019_race.csv')
norm_race_2020 = pd.read_csv('norm_2020_race.csv')
norm_race_2021 = pd.read_csv('norm_2021_race.csv')

# 각 데이터프레임에서 컬럼 이름 변경
norm_race_2017 = norm_race_2017.rename(columns={'State': 'state_name'})
norm_race_2018 = norm_race_2018.rename(columns={'State': 'state_name'})
norm_race_2019 = norm_race_2019.rename(columns={'State': 'state_name'})
norm_race_2020 = norm_race_2020.rename(columns={'State': 'state_name'})
norm_race_2021 = norm_race_2021.rename(columns={'State': 'state_name'})


df2_2017 = pd.merge(norm_race_2017, income_2017, on='state_name')
df2_2018 = pd.merge(norm_race_2018, income_2018, on='state_name')
df2_2019 = pd.merge(norm_race_2019, income_2019, on='state_name')
df2_2020 = pd.merge(norm_race_2020, income_2020, on='state_name')
df2_2021 = pd.merge(norm_race_2021, income_2021, on='state_name')

data_2017 = pd.merge(df1_2017 , norm_race_2017, on='state_name')
data_2018 = pd.merge(df1_2018, norm_race_2018, on='state_name')
data_2019 = pd.merge(df1_2019, norm_race_2019, on='state_name')
data_2020 = pd.merge(df1_2020, norm_race_2020, on='state_name')
data_2021 = pd.merge(df1_2021, norm_race_2021, on='state_name')


### 상관분석 _ 피어슨 상관계수 사용
#### 주별 소득수준과 범죄횟수 간의 상관관계 파악

피어슨 상관 계수는 연속형 변수 간의 상관 관계를 평가하는 데 적합하며, 스피어만 상관 계수는 순위 형태의 변수 또는 비선형 관계를 갖는 변수 간의 상관 관계를 평가하는 데 적합

In [7]:
# 연도별 데이터를 담을 빈 리스트 생성
correlation_results = []

# 연도별로 데이터 분할 및 피어슨 상관 계수 계산
for year in range(2017, 2022):
    # 해당 연도의 데이터프레임 선택
    data = globals()[f"data_{year}"]
    
    # column선택 : state_name, hate_crime_case_normalized, median_income
    selected_data = data[['state_name', 'hate_crime_case_normalized', 'median_income']]
    
    # 피어슨 상관 계수 계산
    pearson_corr = selected_data['hate_crime_case_normalized'].corr(selected_data['median_income'], method='pearson')
    
    # 결과를 딕셔너리로 저장
    correlation_results.append({'Year': year, 'Pearson Correlation Coefficient': pearson_corr})

    # 산점도 출력
    x = data['median_income']
    y = data['hate_crime_case_normalized']

    # 산점도 그리기
    # plt.scatter(x, y)
    # plt.xlabel('Median Income')
    # plt.ylabel('Hate Crime Case Normalized')
    # plt.title('Scatter Plot of Median Income vs. Hate Crime Case Normalized')
    # plt.show()

# 결과 출력
for result in correlation_results:
    print(f"Year: {result['Year']}, Pearson Correlation Coefficient: {result['Pearson Correlation Coefficient']}")

Year: 2017, Pearson Correlation Coefficient: 0.5026505362463362
Year: 2018, Pearson Correlation Coefficient: 0.5460568674434366
Year: 2019, Pearson Correlation Coefficient: 0.6174406329516006
Year: 2020, Pearson Correlation Coefficient: 0.5026584234922596
Year: 2021, Pearson Correlation Coefficient: 0.3288552548021082


In [29]:
import statsmodels.api as sm

# 연도별 데이터 파일명 리스트
datasets = [data_2017, data_2018, data_2019, data_2020, data_2021]

# 다중 회귀 분석을 위한 빈 DataFrame 생성
results_df = pd.DataFrame(columns=['Year', 'R-squared', 'Median Income Coef', 'Black Coef', 'Hispanic Coef', 'Asian Coef', 'Multiple Coef', 'Others Coef'])

# 연도별로 다중 회귀 분석 수행
for year, dataset in zip(range(2017, 2022), datasets):
    # 데이터 로드
    data = dataset

    # 결측값 및 무한값 처리
    data = data.replace([np.inf, -np.inf], np.nan)
    data = data.dropna()  # 결측값이 있는 행 제거

    # 독립 변수(X)와 종속 변수(Y) 설정
    X = data[['median_income', 'Black', 'Hispanic', 'Asian', 'Multiple', 'Others']]  # 독립 변수 열 선택
    y = data['hate_crime_case_normalized']  # 종속 변수 선택

    # 상수(intercept) 추가
    X = sm.add_constant(X)

    # 다중 회귀 모델 피팅
    model = sm.OLS(y, X)
    results = model.fit()

    # 회귀 분석 결과 저장
    results_df.loc[year] = [year, results.rsquared, results.params['median_income'], results.params['Black'], results.params['Hispanic'], results.params['Asian'], results.params['Multiple'], results.params['Others']]

# 결과 출력
print(results_df)


        Year  R-squared  Median Income Coef  Black Coef  Hispanic Coef  \
2017  2017.0   0.744085            0.000002    1.228871      32.475790   
2018  2018.0   0.145714           -0.000204   -0.291139      -9.081578   
2019  2019.0   0.473207            0.000468    1.433336       6.330653   
2020  2020.0   0.811278            0.000475    1.044078       8.403950   
2021  2021.0   0.548291            0.000590    0.725851      50.562129   

      Asian Coef  Multiple Coef  Others Coef  
2017  121.654301       4.241668          0.0  
2018    3.259572      11.062084          0.0  
2019    4.016121      10.874429          0.0  
2020   37.930174      14.209763          0.0  
2021   -5.200683       7.171501          0.0  


## 분산분석(ANOVA)
### 종속변수가 정규분포를 따르지 않기 때문에 분산분석은 활용X
### -> 비모수적인 분석 방법 사용 : e.g., 크루스칼-월리스 검정,Mann-Whitney U (검정그룹간의 중앙값이 서로 다른지를 비교하는 검정), 회귀 분석, 스피어만 상관 계수 등

In [16]:
# import scipy.stats as stats

# datasets = [data_2017, data_2018, data_2019, data_2020, data_2021]

# for i, data in enumerate(datasets):
#     print(f"Year: {2017 + i}")
# # 크루스칼-월리스 검정 수행
#     formula = 'hate_crime_case_normalized ~ state_name'
#     result = stats.kruskal(*[data[data['state_name'] == state]['hate_crime_case_normalized'] for state in data['state_name'].unique()])

#     print(f"p-value: {result.pvalue}")

# print()




---



---



## 1. 가설설정
### "소득 수준과 범죄횟수는 반비례 관계일 것이다."
#### 귀무가설(null_hypothesis) : 소득수준과 범죄횟수 간에 관계가 없다.
#### 대립가설(alternative hypothesis) : 소득수준이 높아질수록 범죄횟수는 낮다.

## 2. 유의수준 설정
#### 0.05로 설정

## 3. 검정통계량 산출

In [10]:
import statsmodels.api as sm

# List of datasets
datasets = [data_2017, data_2018, data_2019, data_2020, data_2021]

# Perform regression analysis for each dataset
for i, data in enumerate(datasets):
    print(f"Dataset: data_{2017 + i}")
    
    # Select relevant columns
    selected_data = data[['state_name', 'hate_crime_case_normalized', 'median_income']]
    
    # Create regression model
    X = selected_data['median_income']
    X = sm.add_constant(X)
    y = selected_data['hate_crime_case_normalized']
    model = sm.OLS(y, X)
    
    # Fit the regression model
    results = model.fit()
    
    # Print p-value
    print("P-value:", results.pvalues['median_income'])
    print()

Dataset: data_2017
P-value: 0.00017068995397843416

Dataset: data_2018
P-value: 4.940820673264162e-05

Dataset: data_2019
P-value: 1.3917103990576532e-06

Dataset: data_2020
P-value: 0.00017064324446323457

Dataset: data_2021
P-value: 0.0184617865690403



######모든 데이터셋의 p-value 값 < 유의수준 0.05
###### ==> 귀무가설 기각. 대립가설 채택 => 즉, 소득수준과 혐오범죄수 간에는 유의미한 음의 상관 관계가 있음을 의미

### 주성분 분석(Principal Component Analysis, PCA)
#### 다중 변수들 간의 관계를 이해하고 차원 축소를 위해 사용되는 분석 방법 예를 들어, 다양한 인종 퍼센트와 소득 수준과의 관계를 파악하기 위해 PCA를 사용하여 데이터의 차원을 축소하고 주성분을 도출할 수 있습니다.

### 다중 선형 회귀 (Multiple Linear Regression)
#### 소득수준과 인종을 동시에 사용하여 혐오범죄수에 미치는 영향을 분석 
#### 소득수준과 인종을 독립 변수로 혐오범죄수를 종속 변수로 설정하여 수행.

- 결정 계수 (R-squared)는 회귀 분석에서 종속 변수의 변동성을 독립 변수로 설명하는 정도를 나타내는 지표
  - R-squared 값은 0에서 1 사이의 값을 가지며, 높을수록 독립 변수가 종속 변수의 변동성을 잘 설명한다는 것을 의미

#### 우선, 독립변수들 간의 다중공선성(Multicollinearity) 판단
#### 회귀분석의 전제 가정 : 독립변수들 간에 상관관계가 높으면 안됨
#### 일반적으로 5 이상의 값이 나타나면 다중공선성의 증거로 간주

In [12]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

datasets = [data_2017, data_2018, data_2019, data_2020, data_2021]

# Perform regression analysis for each dataset
for i, data in enumerate(datasets):
    print(f"Dataset: data_{2017 + i}")

    # 독립 변수만 선택
    X = data[['median_income', 'Black', 'Hispanic', 'Asian', 'Multiple']]

    # VIF 계산
    vif = pd.DataFrame()
    vif["Features"] = X.columns
    vif["VIF"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

    # 결과 출력
    print(vif)


Dataset: data_2017
        Features       VIF
0  median_income  1.985090
1          Black  2.172472
2       Hispanic  2.487377
3          Asian  2.298526
4       Multiple  1.804589
Dataset: data_2018
        Features        VIF
0  median_income   1.380055
1          Black   5.940042
2       Hispanic   6.329694
3          Asian   4.485327
4       Multiple  10.271187
Dataset: data_2019
        Features       VIF
0  median_income  2.051197
1          Black  2.022737
2       Hispanic  1.438537
3          Asian  1.446182
4       Multiple  1.856140
Dataset: data_2020
        Features       VIF
0  median_income  2.371702
1          Black  1.635947
2       Hispanic  1.572015
3          Asian  2.672712
4       Multiple  1.667103
Dataset: data_2021
        Features       VIF
0  median_income  2.681765
1          Black  2.386352
2       Hispanic  3.251136
3          Asian  4.263103
4       Multiple  2.304328


data_2018을 제외하고서는 VIF값이 다 작으므로 회귀분석 가능
- data_2018은 다중공선성을 해결할 필요가 있음 : e.g., 변수선택, 변수변환, PCA, 규제모델
  - 변수선택 : 다중공선성이 강하게 나타나는 변수들을 제거하여 모델에 포함시키기
  - 변수변환 : 변수들을 다른 형태로 변환하여 다중공선성 줄이기(e.g., 표준화, 로그변환 등)
  - PCA : 변수들 간의 선형 종속성을 고려하여 변수들을 새로운 주성분으로 변환
  - 규제모델 : 릿지 or 라쏘 등을 활용하여 계수의 크기를 조절
- 변수변환 선택

In [13]:
from sklearn.preprocessing import StandardScaler

data = data_2018

# 변수 변환을 위해 변수들을 선택
selected_features = ['median_income', 'Black', 'Hispanic', 'Asian', 'Multiple']

# 변수들의 데이터를 표준화
scaler = StandardScaler()
data[selected_features] = scaler.fit_transform(data[selected_features])

data_2018 = data

In [14]:
# data_2018에 대해 다중공선성 판단 재실시
# 독립 변수만 선택
X = data[['median_income', 'Black', 'Hispanic', 'Asian', 'Multiple']]

# VIF 계산
vif = pd.DataFrame()
vif["Features"] = X.columns
vif["VIF"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

# 결과 출력
print(vif)

        Features       VIF
0  median_income  1.262111
1          Black  5.606133
2       Hispanic  5.690731
3          Asian  4.019919
4       Multiple  8.922317


In [15]:
# 다중회귀분석 진
datasets = [data_2017, data_2018, data_2019, data_2020, data_2021]

# Perform regression analysis for each dataset
for i, data in enumerate(datasets):
    print(f"Dataset: data_{2017 + i}")
    
    # 독립 변수와 종속 변수 설정
    X = data[['median_income', 'Black', 'Hispanic', 'Asian', 'Multiple']]
    y = data['hate_crime_case_normalized']

    # 상수 (절편) 추가
    X = sm.add_constant(X)
    
    # 다중 선형 회귀 모델 학습
    model = sm.OLS(y, X).fit()

    # 회귀 분석 결과 출력
    print(model.summary())

Dataset: data_2017
                                OLS Regression Results                                
Dep. Variable:     hate_crime_case_normalized   R-squared:                       0.728
Model:                                    OLS   Adj. R-squared:                  0.698
Method:                         Least Squares   F-statistic:                     24.10
Date:                        Mon, 29 May 2023   Prob (F-statistic):           1.04e-11
Time:                                16:48:18   Log-Likelihood:                -226.90
No. Observations:                          51   AIC:                             465.8
Df Residuals:                              45   BIC:                             477.4
Df Model:                                   5                                         
Covariance Type:                    nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------

R-squared
- data_2017 : 0.728
- data_2018 : 0.393
- data_2019 : 0.892 => 매우 큰 편
- data_2020 : 0.805 => 매우 큰 편
- data_2021 : 0.558


F-statistic
- data_2017 : 24.10
- data_2018 : 5.574
- data_2019 : 74.34
- data_2020 : 37.26
- data_2021 : 11.38

Cond.No.(=다중공선성을 의미) 값이 대체로 다 크게 나타남 + Strong Multicollineartiy가 존재할 수 있다고 경고 제시 -> 안 하는 게 나을 듯...

In [19]:
import scipy.stats as stats

income_threshold = 50000

high_income = data_2017[data_2017['median_income'] >= income_threshold]['hate_crime_case_normalized']
low_income = data_2017[data_2017['median_income'] < income_threshold]['hate_crime_case_normalized']
result = stats.ttest_ind(high_income, low_income)
print(f"p-value: {result.pvalue}")


p-value: 0.618041703286677


In [23]:
# t-test
# 높은 소득수준 지역과 낮은 소득수준 그룹 간의 범죄횟수 평균 차이 검정
import scipy.stats as stats

def perform_t_test(data, income_threshold):
    high_income = data[data['median_income'] >= income_threshold]['hate_crime_case_normalized']
    low_income = data[data['median_income'] < income_threshold]['hate_crime_case_normalized']
    result = stats.ttest_ind(high_income, low_income)
    return result

income_threshold = 150000  # 소득수준을 기준으로 높은 소득수준과 낮은 소득수준을 구분하기 위한 임계값 설정

datasets = [data_2018, data_2019, data_2020, data_2021]

for year, dataset in enumerate(datasets):
    print(f"Dataset: data_{2018 + year}")
    result = perform_t_test(dataset, income_threshold)
    print(f"p-value: {result.pvalue}")



Dataset: data_2018
p-value: nan
Dataset: data_2019
p-value: nan
Dataset: data_2020
p-value: nan
Dataset: data_2021
p-value: nan


nan 출력 이유 : 해당 그룹의 데이터 부족 , 분산이 없어서 t-test수행 불가능 

=> 소득 수준을 나누는 기준을 낮춰봤음에도 불구하고 안 나옴

전처리 과정을 다시 할 필요가 있음...-> 근데, how?

## 4. 기각/채택 판단

- 회귀 분석에 대한 결과

Dataset: data_2017
P-value: 0.00017068995397843416

Dataset: data_2018
P-value: 4.940820673264162e-05

Dataset: data_2019
P-value: 1.3917103990576532e-06

Dataset: data_2020
P-value: 0.00017064324446323457

Dataset: data_2021
P-value: 0.0184617865690403



- 피어슨 상관계수에 대한 결과 

Year: 2017, Pearson Correlation Coefficient: 0.5026505362463362

Year: 2018, Pearson Correlation Coefficient: 0.5460568674434366

Year: 2019, Pearson Correlation Coefficient: 0.6174406329516006

Year: 2020, Pearson Correlation Coefficient: 0.5026584234922596

Year: 2021, Pearson Correlation Coefficient: 0.3288552548021082



---



---



- 회귀 분석 결과

모든 데이터셋에서 p-value가 유의수준인 0.05보다 많이 작게 나옴.
이는 소득수준과 범죄 횟수 간에 통계적으로 유의한 관계가 있음을 나타냄.
따라서, 회귀 분석 결과는 귀무가설을 기각하고 대립가설을 채택가능.

- 피어슨 상관계수 결과

모든 데이터셋에서 피어슨 상관계수가 양수로 나타남
상관계수 값이 0에 가까울수록 관련성이 약하고, 1에 가까울수록 관련성이 강함.
모든 연도에서 상관계수가 양수이며, 일부 연도에서는 0.5 이상의 상관계수를 보였음.
이는 소득수준과 범죄 횟수 간에 양적인 관련이 있음을 시사함.
따라서, 피어슨 상관계수 결과도 가설을 지지

## 결과
결과적으로, 회귀 분석과 피어슨 상관계수 결과 모두 소득수준과 범죄 횟수 간의 관련성을 지지하므로 가설로 내세운 "소득수준이 높을수록 범죄 횟수가 적을 것이다"는 채택될 수 있음.